482p3/usr.bin/cksum/cksum.1
.\"	$NetBSD: cksum.1,v 1.44 2012/06/25 02:32:12 riastradh Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)cksum.1	8.2 (Berkeley) 4/28/95
.\"
.Dd June 24, 2012
.Dt CKSUM 1
.Os
.Sh NAME
.Nm cksum ,
.Nm md2 ,
.Nm md4 ,
.Nm md5 ,
.Nm rmd160 ,
.Nm sha1 ,
.Nm sum
.Nd display file checksums and block counts
.Sh SYNOPSIS
.Nm cksum
.Op Fl n
.Op Fl a Ar algorithm Oo Fl ptx Oc Oo Fl s Ar string Oc
.Op Fl o Ar 1 Ns | Ns Ar 2
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Nm sum
.Op Fl n
.Op Fl a Ar algorithm Oo Fl ptx Oc Oo Fl s Ar string Oc
.Op Fl o Ar 1 Ns | Ns Ar 2
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Nm md2
.Op Fl nptx
.Op Fl s Ar string
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Nm md4
.Op Fl nptx
.Op Fl s Ar string
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Nm md5
.Op Fl nptx
.Op Fl s Ar string
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Nm rmd160
.Op Fl nptx
.Op Fl s Ar string
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Nm sha1
.Op Fl nptx
.Op Fl s Ar string
.Op Ar Li \&| Fl c Oo Fl w Oc Oo Ar sumfile Oc
.Sh DESCRIPTION
The
.Nm
utility writes to the standard output three whitespace separated
fields for each input file.
These fields are a checksum
.Tn CRC ,
the total number of octets in the file and the file name.
If no file name is specified, the standard input is used and no file name
is written.
.Pp
The
.Nm sum
utility is identical to the
.Nm
utility, except that it defaults to using historic algorithm 1, as
described below.
It is provided for compatibility only.
.Pp
The
.Nm md2 ,
.Nm md4 ,
.Nm md5 ,
.Nm sha1 ,
and
.Nm rmd160
utilities compute cryptographic hash functions, and write to standard
output the hexadecimal representation of the hash of their input.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl a Ar algorithm
When invoked as
.Nm cksum ,
use the specified
.Ar algorithm .
Valid algorithms are:
.Bl -column -offset indent ".Sy Algorithm" ".Sy Bits" ".Sy Description"
.It Sy Algorithm Ta Sy Bits Ta Sy Description
.It Li CRC Ta 32 Ta Default CRC algorithm
.It Li MD2 Ta 128 Ta MD2, per Li RFC1319
.It Li MD4 Ta 128 Ta MD4, per Li RFC1320
.It Li MD5 Ta 128 Ta MD5, per Li RFC1321
.It Li RMD160 Ta 160 Ta RIPEMD-160
.It Li SHA1 Ta 160 Ta SHA-1, per Li FIPS PUB 180-1
.It Li SHA256 Ta 256 Ta SHA-2
.It Li SHA384 Ta 384 Ta SHA-2
.It Li SHA512 Ta 512 Ta SHA-2
.It Li old1 Ta 16 Ta Algorithm 1, per Fl o Ar 1
.It Li old2 Ta 16 Ta Algorithm 2, per Fl o Ar 2
.El
.It Fl c Op Ar sumfile
Verify (check) files against a list of checksums.
The list is read from
.Ar sumfile ,
or from stdin if no filename is given.
E.g. first run
.Dl Ic md5 *.tgz \*[Gt] MD5
.Dl Ic sha1 *.tgz \*[Gt] SHA1
to generate a list of MD5 checksums in
.Pa MD5 ,
then use the following command to verify them:
.Dl Ic cat MD5 SHA1 | cksum -c
If an error is found during checksum verification, an error
message is printed, and the program returns an error code of 1.
.It Fl o
Use historic algorithms instead of the (superior) default one.
.Pp
Algorithm 1 is the algorithm used by historic
.Bx
systems as the
.Xr sum 1
algorithm and by historic
.At V
systems as the
.Xr sum 1
algorithm when using the
.Fl r
option.
This is a 16-bit checksum, with a right rotation before each addition;
overflow is discarded.
.Pp
Algorithm 2 is the algorithm used by historic
.At V
systems as the
default
.Xr sum 1
algorithm.
This is a 32-bit checksum, and is defined as follows:
.Bd -unfilled -offset indent
s = sum of all bytes;
r = s % 2^16 + (s % 2^32) / 2^16;
cksum = (r % 2^16) + r / 2^16;
.Ed
.Pp
Both algorithm 1 and 2 write to the standard output the same fields as
the default algorithm except that the size of the file in bytes is
replaced with the size of the file in blocks.
For historic reasons, the block size is 1024 for algorithm 1 and 512
for algorithm 2.
Partial blocks are rounded up.
.It Fl w
Print warnings about malformed checksum files when verifying
checksums with
.Fl c .
.El
.Pp
The following options apply only when using the one of the message
digest algorithms:
.Bl -tag -width indent
.It Fl n
Print the hash and the filename in the normal sum output form, with
the hash at the left and the filename following on the right.
.It Fl p
Echo input from standard input to standard output, and append the
selected message digest.
.It Fl s Ar string
Print the hash of the given string
.Ar string .
.It Fl t
Run a built-in message digest time trial.
.It Fl x
Run a built-in message digest test script.
The tests that are run
are supposed to encompass all the various tests in the suites that
accompany the algorithms' descriptions with the exception of the
last test for the SHA-1 algorithm and the RIPEMD-160 algorithm.
The
last test for these is one million copies of the lower letter a.
.El
.Pp
The default
.Tn CRC
used is based on the polynomial used for
.Tn CRC
error checking
in the networking standard
.St -iso8802-3 .
The
.Tn CRC
checksum encoding is defined by the generating polynomial:
.Pp
.Bd -unfilled -offset indent
G(x) = x^32 + x^26 + x^23 + x^22 + x^16 + x^12 +
     x^11 + x^10 + x^8 + x^7 + x^5 + x^4 + x^2 + x + 1
.Ed
.Pp
Mathematically, the
.Tn CRC
value corresponding to a given file is defined by
the following procedure:
.Bd -filled -offset indent
The
.Ar n
bits to be evaluated are considered to be the coefficients of a mod 2
polynomial M(x) of degree
.Ar n Ns \-1 .
These
.Ar n
bits are the bits from the file, with the most significant bit being the most
significant bit of the first octet of the file and the last bit being the least
significant bit of the last octet, padded with zero bits (if necessary) to
achieve an integral number of octets, followed by one or more octets
representing the length of the file as a binary value, least significant octet
first.
The smallest number of octets capable of representing this integer are used.
.Pp
M(x) is multiplied by x^32 (i.e., shifted left 32 bits) and divided by
G(x) using mod 2 division, producing a remainder R(x) of degree \*[Le] 31.
.Pp
The coefficients of R(x) are considered to be a 32-bit sequence.
.Pp
The bit sequence is complemented and the result is the CRC.
.Ed
.Pp
The
.Nm
and
.Nm sum
utilities exit 0 on success, and \*[Gt]0 if an error occurs.
.Sh SEE ALSO
.Xr openssl 1 ,
.Xr mtree 8
.Pp
The default calculation is identical to that given in pseudo-code
in the following
.Tn ACM
article.
.Rs
.%T "Computation of Cyclic Redundancy Checks Via Table Lookup"
.%A Dilip V. Sarwate
.%J "Communications of the \*(tNACM\*(sP"
.%D "August 1988"
.Re
.Rs
.%A R. Rivest
.%T The MD2 Message-Digest Algorithm
.%O RFC 1319
.Re
.Rs
.%A R. Rivest
.%T The MD4 Message-Digest Algorithm
.%O RFC 1186 and RFC 1320
.Re
.Rs
.%A R. Rivest
.%T The MD5 Message-Digest Algorithm
.%O RFC 1321
.Re
.Rs
.%A U.S. DOC/NIST
.%T Secure Hash Standard
.%O FIPS PUB 180-1
.Re
.Sh STANDARDS
The
.Nm
utility is expected to conform to
.St -p1003.1-2004 .
.Sh HISTORY
The
.Nm
utility appeared in
.Bx 4.4 .
.Nm md5
was added in
.Nx 1.3 .
The functionality for
.Nm md2 ,
.Nm md4 ,
.Nm sha1 ,
and
.Nm rmd160
was added in
.Nx 1.6 .
Support for the SHA-2 algorithms
.Po
.Li SHA256 ,
.Li SHA384 ,
and
.Li SHA512
.Pc
was added in
.Nx 3.0 .
The functionality to verify checksum stored in a file
.Pq Fl c
first appeared in
.Nx 4.0 .
.\" .Pp
.\" The
.\" .Nm sum
.\" utility appeared in
.\" .Bx ?.?
.\" and
.\" .At V .


p3team-05/usr.bin/cksum/cksum.c
/*	$NetBSD: cksum.c,v 1.45 2011/08/29 14:12:29 joerg Exp $	*/

/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * James W. Williams of NASA Goddard Space Flight Center.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*-
 * Copyright (c) 1997 Jason R. Thorpe.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * James W. Williams of NASA Goddard Space Flight Center.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if HAVE_NBTOOL_CONFIG_H
#include "nbtool_config.h"
#endif

#include <sys/cdefs.h>
#if defined(__COPYRIGHT) && !defined(lint)
__COPYRIGHT("@(#) Copyright (c) 1991, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#if defined(__RCSID) && !defined(lint)
#if 0
static char sccsid[] = "@(#)cksum.c	8.2 (Berkeley) 4/28/95";
#endif
__RCSID("$NetBSD: cksum.c,v 1.45 2011/08/29 14:12:29 joerg Exp $");
#endif /* not lint */

#include <sys/types.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <locale.h>
#include <md2.h>
#include <md4.h>
#include <md5.h>
#include <rmd160.h>
#include <sha1.h>
#include <sha2.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "extern.h"

typedef char *(*_filefunc)(const char *, char *);

const struct hash {
	const char *progname;
	const char *hashname;
	void (*stringfunc)(const char *);
	void (*timetrialfunc)(void);
	void (*testsuitefunc)(void);
	void (*filterfunc)(int);
	char *(*filefunc)(const char *, char *);
} hashes[] = {
	{ "md2", "MD2",
	  MD2String, MD2TimeTrial, MD2TestSuite,
	  MD2Filter, MD2File },
	{ "md4", "MD4",
	  MD4String, MD4TimeTrial, MD4TestSuite,
	  MD4Filter, MD4File },
	{ "md5", "MD5",
	  MD5String, MD5TimeTrial, MD5TestSuite,
	  MD5Filter, MD5File },
	{ "rmd160", "RMD160",
	  RMD160String, RMD160TimeTrial, RMD160TestSuite,
	  RMD160Filter, (_filefunc) RMD160File },
	{ "sha1", "SHA1",
	  SHA1String, SHA1TimeTrial, SHA1TestSuite,
	  SHA1Filter, (_filefunc) SHA1File },
	{ "sha256", "SHA256",
	  SHA256_String, SHA256_TimeTrial, SHA256_TestSuite,
	  SHA256_Filter, (_filefunc) SHA256_File },
	{ "sha384", "SHA384",
	  SHA384_String, SHA384_TimeTrial, SHA384_TestSuite,
	  SHA384_Filter, (_filefunc) SHA384_File },
	{ "sha512", "SHA512",
	  SHA512_String, SHA512_TimeTrial, SHA512_TestSuite,
	  SHA512_Filter, (_filefunc) SHA512_File },
	{ .progname = NULL, },
};

static int	hash_digest_file(char *, const struct hash *, int);
__dead static void	requirehash(const char *);
__dead static void	usage(void);

int
main(int argc, char **argv)
{
	int ch, fd, rval, dosum, pflag, nohashstdin;
	u_int32_t val;
	off_t len;
	char *fn;
	const char *progname;
	int (*cfncn) (int, u_int32_t *, off_t *);
	void (*pfncn) (char *, u_int32_t, off_t);
	const struct hash *hash;
	int normal, i, check_warn, do_check;

	cfncn = NULL;
	pfncn = NULL;
	dosum = pflag = nohashstdin = 0;
	normal = 0;
	check_warn = 0;
	do_check = 0;

	setlocale(LC_ALL, "");

	progname = getprogname();

	for (hash = hashes; hash->hashname != NULL; hash++)
		if (strcmp(progname, hash->progname) == 0)
			break;

	if (hash->hashname == NULL) {
		hash = NULL;

		if (!strcmp(progname, "sum")) {
			dosum = 1;
			cfncn = csum1;
			pfncn = psum1;
		} else {
			cfncn = crc;
			pfncn = pcrc;
		}
	}

	while ((ch = getopt(argc, argv, "a:cno:ps:twx")) != -1)
		switch(ch) {
		case 'a':
			if (hash) {
				warnx("illegal use of -a option\n");
				usage();
			}
			i = 0;
			while (hashes[i].hashname != NULL) {
				if (!strcasecmp(hashes[i].hashname, optarg)) {
					hash = &hashes[i];
					break;
				}
				i++;
			}
			if (hash == NULL) {
				if (!strcasecmp(optarg, "old1")) {
					cfncn = csum1;
					pfncn = psum1;
				} else if (!strcasecmp(optarg, "old2")) {
					cfncn = csum2;
					pfncn = psum2;
				} else if (!strcasecmp(optarg, "crc")) {
					cfncn = crc;
					pfncn = pcrc;
				} else {
					warnx("illegal argument to -a option");
					usage();
				}
			}
			break;
		case 'c':
			do_check = 1;
			break;
		case 'n':
			normal = 1;
			break;
		case 'o':
			if (hash) {
				warnx("%s mutually exclusive with sum",
				      hash->hashname);
				usage();
			}
			if (!strcmp(optarg, "1")) {
				cfncn = csum1;
				pfncn = psum1;
			} else if (!strcmp(optarg, "2")) {
				cfncn = csum2;
				pfncn = psum2;
			} else {
				warnx("illegal argument to -o option");
				usage();
			}
			break;
		case 'p':
			if (hash == NULL)
				requirehash("-p");
			pflag = 1;
			break;
		case 's':
			if (hash == NULL)
				requirehash("-s");
			nohashstdin = 1;
			hash->stringfunc(optarg);
			break;
		case 't':
			if (hash == NULL)
				requirehash("-t");
			nohashstdin = 1;
			hash->timetrialfunc();
			break;
		case 'w':
			check_warn = 1;
			break;
		case 'x':
			if (hash == NULL)
				requirehash("-x");
			nohashstdin = 1;
			hash->testsuitefunc();
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (do_check) {
		/*
		 * Verify checksums
		 */
		FILE *f;
		char buf[BUFSIZ];
		char *s, *p_filename, *p_cksum;
		int l_filename, l_cksum;
		char filename[BUFSIZ];
		char cksum[BUFSIZ];
		int ok,cnt,badcnt;

		rval = 0;
		cnt = badcnt = 0;

		if (argc == 0) {
			f = fdopen(STDIN_FILENO, "r");
		} else {
			f = fopen(argv[0], "r");
		}
		if (f == NULL)
			err(1, "Cannot read %s",
			    argc>0?argv[0]:"stdin");
		
		while(fgets(buf, sizeof(buf), f) != NULL) {
			s=strrchr(buf, '\n');
			if (s)
				*s = '\0';

			p_cksum = p_filename = NULL;

			p_filename = strchr(buf, '(');
			if (p_filename) {
				/*
				 * Assume 'normal' output if there's a '('
				 */
				p_filename += 1;
				normal = 0;

				p_cksum = strrchr(p_filename, ')');
				if (p_cksum == NULL) {
					if (check_warn)
						warnx("bogus format: %s. "
						      "Skipping...",
						      buf);
					rval = 1;
					continue;
				}
 				p_cksum += 4;

				l_cksum = strlen(p_cksum);
				l_filename = p_cksum - p_filename - 4;
					
				/* Sanity check, and find proper hash if
				 * it's not the same as the current program
				 */
				if (hash == NULL ||
				    strncmp(buf, hash->hashname,
					    strlen(hash->hashname)) != 0) {
					/*
					 * Search proper hash
					 */
					const struct hash *nhash;
					
					for (nhash = hashes ;
					     nhash->hashname != NULL;
					     nhash++)
						if (strncmp(buf,
							    nhash->hashname,
							    strlen(nhash->hashname)) == 0)
							break;
					
					
					if (nhash->hashname == NULL) {
						if (check_warn)
							warnx("unknown hash: %s",
							      buf);
						rval = 1;
						continue;
					} else {
						hash = nhash;
					}
				}

			} else {
				if (hash) {
					int nspaces;

					/*
					 * 'normal' output, no (ck)sum
					 */
					normal = 1;
					nspaces = 1;
					
					p_cksum = buf;
					p_filename = strchr(buf, ' ');
					if (p_filename == NULL) {
						if (check_warn)
							warnx("no filename in %s? "
							      "Skipping...", buf);
						rval = 1;
						continue;
					}
					while (isspace((int)*++p_filename))
						nspaces++;
					l_filename = strlen(p_filename);
					l_cksum = p_filename - buf - nspaces;
				} else {
					/*
					 * sum/cksum output format
					 */
					p_cksum = buf;
					s=strchr(p_cksum, ' ');
					if (s == NULL) {
						if (check_warn)
							warnx("bogus format: %s."
							      " Skipping...",
							      buf);
						rval = 1;
						continue;
					}
					l_cksum = s - p_cksum;

					p_filename = strrchr(buf, ' ');
					if (p_filename == NULL) {
						if (check_warn)
							warnx("no filename in %s?"
							      " Skipping...",
							      buf);
						rval = 1;
						continue;
					}
					p_filename++;
					l_filename = strlen(p_filename);
				}
			}

			strlcpy(filename, p_filename, l_filename+1);
			strlcpy(cksum, p_cksum, l_cksum+1);

			if (hash) {
				if (access(filename, R_OK) == 0
				    && strcmp(cksum, hash->filefunc(filename, NULL)) == 0)
					ok = 1;
				else
					ok = 0;
			} else {
				if ((fd = open(filename, O_RDONLY, 0)) < 0) {
					if (check_warn)
						warn("%s", filename);
					rval = 1;
					ok = 0;
				} else {
					if (cfncn(fd, &val, &len)) 
						ok = 0;
					else {
						u_int32_t should_val;
						
						should_val =
						  strtoul(cksum, NULL, 10);
						if (val == should_val)
							ok = 1;
						else
							ok = 0;
					}
					close(fd);
				}
			}

			if (! ok) {
				if (hash)
					printf("(%s) ", hash->hashname);
				printf("%s: FAILED\n", filename);
				badcnt++;
			}
			cnt++;

		}
		fclose(f);

		if (badcnt > 0) 
			rval = 1;
		
	} else {
		/*
		 * Calculate checksums
		 */

		fd = STDIN_FILENO;
		fn = NULL;
		rval = 0;
		do {
			if (*argv) {
				fn = *argv++;
				if (hash != NULL) {
					if (hash_digest_file(fn, hash, normal)) {
						warn("%s", fn);
						rval = 1;
					}
					continue;
				}
				if ((fd = open(fn, O_RDONLY, 0)) < 0) {
					warn("%s", fn);
					rval = 1;
					continue;
				}
			} else if (hash && !nohashstdin) {
				hash->filterfunc(pflag);
			}
			
			if (hash == NULL) {
				if (cfncn(fd, &val, &len)) {
					warn("%s", fn ? fn : "stdin");
					rval = 1;
				} else
					pfncn(fn, val, len);
				(void)close(fd);
			}
		} while (*argv);
	}
	exit(rval);
}

static int
hash_digest_file(char *fn, const struct hash *hash, int normal)
{
	char *cp;

	cp = hash->filefunc(fn, NULL);
	if (cp == NULL)
		return 1;

	if (normal)
		printf("%s %s\n", cp, fn);
	else
		printf("%s (%s) = %s\n", hash->hashname, fn, cp);

	free(cp);

	return 0;
}

static void
requirehash(const char *flg)
{
	warnx("%s flag requires `-a algorithm'", flg);
	usage();
}

static void
usage(void)
{
	const char fileargs[] = "[file ... | -c [-w] [sumfile]]";
	const char sumargs[] = "[-n] [-a algorithm [-ptx] [-s string]] [-o 1|2]";
	const char hashargs[] = "[-nptx] [-s string]";

	(void)fprintf(stderr, "usage: cksum %s\n             %s\n",
	    sumargs, fileargs);
	(void)fprintf(stderr, "       sum %s\n           %s\n",
	    sumargs, fileargs);
	(void)fprintf(stderr, "       md2 %s %s\n", hashargs, fileargs);
	(void)fprintf(stderr, "       md4 %s %s\n", hashargs, fileargs);
	(void)fprintf(stderr, "       md5 %s %s\n", hashargs, fileargs);
	(void)fprintf(stderr, "       rmd160 %s %s\n", hashargs, fileargs);
	(void)fprintf(stderr, "       sha1 %s %s\n", hashargs, fileargs);
	exit(1);
}
